Archaius2 bootstrap cleanup

elandau opened the pull request:
Introduce ArchaiusConfig as the top level opinionated archaius override structure with a Configurator and Consumer based pattern of customizing the configurator.  Also introduce a MatchingDecoder as a mechanism to extend the default decoder for custom types.

This PR solves a few specific problems
1.  Simplified extending via the set bindings to Consumer<Configurator> to override cascade strategy or application name without the need for a module override
2.  Ability to extend the MatchingDecoder
3.  Enables moving more code into ArchaiusConfig instead of having complex DI framework integrations
4.

Breaking changes
1.  @Raw is no longer supported and will break nf-archaius2's use of it to initialize persisted configurationclient
2.

elandau committed:
Introduce ArchaiusConfig as the top level opinionated archaius override structure with a Configurator and Consumer based pattern of customizing the configurator.  Also introduce a MatchingDecoder as a mechanism to extend the default decoder for custom types.

elandau committed:
Remove @Raw since it's fully broken now and replace with methods that take a Function<Config, Config> where the argument is the raw configuration so far
ConfigurationInjectingListener now calls into ArchaiusConfig

elandau commented on archaius2-api/src/main/java/com/netflix/archaius/api/ArchaiusConfig.java:
```diff
@@ -0,0 +1,135 @@
+/**
+ * Capture the entire archaius configuration. 
+ */
+public interface ArchaiusConfig {
```
Any thoughts on the name of this interface?  Should this be ArchaiusManager?

twicksell replied:
Can we call this anything else? ArchaiusBuilder, ArchaiusInitializer? Just don't want yet another "Config"

twicksell commented on archaius2-core/src/main/java/com/netflix/archaius/DefaultArchaiusConfigurator.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Override
+    public Configurator addDecoder(MatchingDecoder decoder) {
```
for all of these, withThing, not addThing

elandau committed:
Fix generics bug

elandau committed:
.

elandau committed:
ConfigManager implementation complete

elandau committed:
Remove incorrect argument check

twicksell commented on archaius2-guice/src/main/java/com/netflix/archaius/guice/InternalArchaiusModule.java:
```diff
@@ -68,159 +39,40 @@ protected void configure() {
-    @Raw
-    CompositeConfig getRawCompositeConfig() throws Exception {
-        return new DefaultCompositeConfig();
+    ConfigManager getConfigManager(DefaultConfigManager.Builder builder) {
```
Make a duplicate binding with the @Raw annotation to keep some backwards compatibility

elandau committed:
Remove resourceName from Key and make it an argument to all addResourceToLayer methods

elandau committed:
Re-add LibrariesLayer with a minimal bridge for backwards compatibility

twicksell commented on archaius2-api/src/main/java/com/netflix/archaius/api/ConfigManager.java:
```diff
@@ -0,0 +1,131 @@
+     * @param layer
+     * @param resourceName
+     */
+    void addResourceToLayer(Key layer, String resourceName);
```
How does adding a resource at a layer affect interpolated config values? Do I expect them to change?

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/ConfigManager.java:
```diff
@@ -0,0 +1,131 @@
+    void addResourceToLayer(Key layer, String resourceName);
+
+    /**
+     * Add a named resource to a configuration layer.  The resourceName identifies a base file name
```
resource != file.  Can we use actual resources rather than assuming file system?

twicksell commented on archaius2-api/src/main/java/com/netflix/archaius/api/inject/DefaultLayer.java:
```diff
@@ -10,6 +10,7 @@
 @Qualifier
 @Target({ElementType.TYPE, ElementType.FIELD, ElementType.PARAMETER, ElementType.METHOD})
 @Retention(RetentionPolicy.RUNTIME)
+@Deprecated
```
Not sure how many people have coded to this, but we should probably have at least a pointer to something saying what to use instead.

twicksell commented on archaius2-archaius1-bridge/src/main/java/com/netflix/archaius/bridge/AbstractConfigurationBridge.java:
```diff
@@ -45,13 +43,11 @@
-            final Config config, 
-            @LibrariesLayer CompositeConfig libraries, 
+            final ConfigManager config,
             @RuntimeLayer SettableConfig settable, 
```
Why would you not also be getting rid of @RuntimeLayer SettableConfig here?

twicksell commented on archaius2-archaius1-bridge/src/main/java/com/netflix/archaius/bridge/AbstractConfigurationBridge.java:
```diff
@@ -134,8 +133,10 @@ public Configuration getConfiguration(int index) {
 
     @Override
     public Configuration removeConfiguration(String name) {
```
Really should've returned optional here.

twicksell commented on archaius2-core/src/main/java/com/netflix/archaius/DefaultConfigManager.java:
```diff
@@ -0,0 +1,482 @@
+         * @return Chainable builder
+         */
+        public Builder withCascadeStrategy(CascadeStrategy strategy) {
+            Preconditions.checkState(rootConfig != null, "Builder already built");
```
Why are these preconditions necessary? If build doesn't return a Builder, how are you going to create this problem?

elandau committed:
Re-add support for @Raw

elandau committed:
Introduce PropertySource api

elandau committed:
Add API to get subset of properties

elandau committed:
Move listener support into PropertySource
Prep for resolving complex types

elandau committed:
.

elandau committed:
First pass at support for proxies and collections

elandau committed:
.

elandau committed:
Remove subset from PropertySource API

elandau committed:
Complete MutablePropertySource

elandau committed:
Switch to using creator instead of resolvers

elandau committed:
Make interpolation lazy while iterating all properties

elandau committed:
Add support for DefaultValue

elandau committed:
fix bugs with PropertySource value supplier

elandau committed:
Simplify registering PropertySource listeners by adding garbage collection

elandau committed:
Replace PropertySource::forEach with PropertySource::Stream of Map.Entry<String, Supplier<Object>>

elandau committed:
Move code from ResolvingPropertySource and InterpolatingPropertySource into Configuration

elandau committed:
Add the Configuration interface to the archaius2-api subproject
Add common streaming use cases to PropertySource including: namespaces, flattened, children
API to add customer creators to CreatorFactoryBuilder

elandau committed:
misc cleanup

elandau committed:
Misc code cleanup
Support for loading from YAML and Property files

elandau committed:
Add unit tests

elandau committed:
Rename CreatorFactory to CreatorRegistry
Simplify API for creating proxies by moving the Collector creation to a Creator#toCollector method

elandau committed:
Replace 'Creator' API with 'Resolver'

elandau committed:
Update PropertySourceBasedConfiguration to use resolvers

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/ConfigManager.java:
```diff
@@ -0,0 +1,71 @@
+     */
+    Optional<Config> removeConfig(OrderedKey layer, String name);
+
+    Iterable<String> getConfigNames();
```
does not support layers well here; should be able to specify the layer name as an argument

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/OrderedKey.java:
```diff
@@ -0,0 +1,63 @@
+        return true;
+    }
+
+    @Override
```
do this during construction

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/PropertyNode.java:
```diff
@@ -0,0 +1,16 @@
+public interface PropertyNode {
+    Optional<?> getValue();
+    
+    default PropertyNode getNode(String key) {
```
getNode->getChild?  Stream<PropertyNode> children()? should there be a 'getParent' as well?

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/PropertySource.java:
```diff
@@ -0,0 +1,142 @@
+     * @param key
+     * @return Value for the property 
+     */
+    Optional<?> getProperty(String key);
```
PropertyNode throws an IllegalStateException for an invalid key, but PropertySource returns an Optional<?>.  seems like we need just one strategy

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/PropertySource.java:
```diff
@@ -0,0 +1,142 @@
+    /**
+     * @return Stream of all entries of this PropertySource
+     */
+    Stream<Map.Entry<String, Object>> stream();
```
as we discussed, PropertyNode might work better for the stream type

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/PropertySource.java:
```diff
@@ -0,0 +1,142 @@
+import java.util.stream.Stream;
+
+/**
+ * Contract for a source of raw properties.  A PropertySource may be associated with a single
```
need to introduce namespaces concept here

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/PropertySource.java:
```diff
@@ -0,0 +1,142 @@
+     * @param namespaces
+     * @return Stream of unique entries from first seen namespace
+     */
+    default Stream<Map.Entry<String, Object>> fallbacks(String... namespaces) {
```
unclear why this code needs to exist (and method name is rather opinionated); consolidate with the other 'stream' method and call it 'filter'

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/exceptions/ConfigException.java:
```diff
@@ -15,7 +15,7 @@
 package com.netflix.archaius.api.exceptions;
 
-public class ConfigException extends Exception {
+public class ConfigException extends RuntimeException {
```
ConfigException isn't very specific, essentially 'GeneralArchaiusFault'.  How about subtypes for ValueConversionException, DuplicateKeyException, ConfigLoaderException, etc.

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/Configuration.java:
```diff
@@ -0,0 +1,39 @@
+
+    public boolean isEmpty();
+
+    PS getPropertySource();
```
why is this exposed?

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/OrderedKey.java:
```diff
@@ -0,0 +1,63 @@
+ * over higher orders.  Within a layer configurations are prioritized by
+ * insertion order (or reversed if 'reversed=true')
+ */
+public final class OrderedKey {
```
Please rename to 'Layer'

tcellucci commented on archaius2-api/src/main/java/com/netflix/archaius/api/PropertySource.java:
```diff
@@ -0,0 +1,142 @@
+     * @param listener
+     * @return Runnable that unregisters the listener when its run() method is called. 
+     */
+    default Runnable addListener(Consumer<PropertySource> listener) { return () -> {}; }
```
return Closeable / AutoCloseable

tcellucci commented:
there's a problem with the term 'resourceName' used in Archaius; that term has been used for ages by the JDK itself as a key for classpath resources.  Archaius describes it as this:
  "The resourceName identifies a base file name without the extension (ex. 'application')"
That's really more along the lines of what the jdk vernacular refers to as a 'bundle', where multiple resources are loaded with some kind of a precedence / override strategy (e.g., ResourceBundle in java's i18n).  

I think Archaius would do well to refer to our 'base file names' as property or config bundles, as it better conveys that multiple property-bearing artifacts are combined into a single config.

elandau committed:
Address some of Tom's comments

elandau committed:
YAML conditional scoping based on any property value

elandau committed:
.

elandau committed:
misc cleanup

elandau committed:
Move all new classes to com.netflix.config package

twicksell commented on archaius2-api/src/main/java/com/netflix/config/api/Configuration.java:
```diff
@@ -0,0 +1,63 @@
+     * @param type
+     * @return
+     */
+    public Optional<Object> get(String key, Type type);
```
Return type probably shouldn't be object. Type should probably be ParameterizedType

elandau committed:
.

elandau committed:
no message

elandau committed:
.

elandau committed:
no message

elandau committed:
no message

elandau committed:
WIP
